File: hall/prediction_monitor.py
"""Read-only, fixed-window monitoring of archived forecasts; never a live gate.

Alerts are descriptive review prompts, not significance tests or model adoption.
Use a common subject panel and fully observed days to avoid outcome cherry-picking.
"""
from __future__ import annotations

from collections import Counter, defaultdict
from datetime import date, datetime, timedelta
from statistics import mean

from hall.prediction_log import JST
from hall.prediction_benchmark import MODELS
from hall.probability_validation import read_samples, stamp

PROTOCOL = 'forecast-monitor-v1'
SPEC = {
    'protocol': PROTOCOL, 'baseline_days': 56, 'recent_days': 28,
    'minimum_baseline_days': 28, 'minimum_baseline_weeks': 6,
    'minimum_recent_days': 14, 'minimum_recent_weeks': 3,
    'minimum_half_days': 7, 'minimum_resolution': .8,
    'minimum_panel_share': .8, 'freshness_days': 7,
    'brier_increase': .05, 'mae_increase_coins': 100, 'mae_increase_ratio': .2,
    'outcome_shift_coins': 500, 'positive_shift_points': 20,
    'weighting': '両期間共通の対象全件が保存・照合された日のみ。対象平均→日を等重み',
    'label': '公開差枚が0枚超（±0は非プラス）。高設定・本人勝率とは別',
    'live_applied': False,
}


def summarize_window(rows, panel, *, start, end):
    rows = [r for r in rows if start.isoformat() <= r['target_date'] <= end.isoformat()]
    selected = [r for r in rows if tuple(r['key']) in panel]
    by_day = defaultdict(list)
    for r in selected:
        by_day[r['target_date']].append(r)
    daily = []
    for day, members in sorted(by_day.items()):
        if {tuple(r['key']) for r in members} != panel or any(r['actual'] is None for r in members):
            continue
        daily.append({'date': day, 'subjects': len(members),
            'brier': mean((r['probability'] - int(r['actual'] > 0))**2 for r in members),
            'mae_coins': mean(abs(r['projected'] - r['actual']) for r in members),
            'avg_diff_coins': mean(r['actual'] for r in members),
            'positive_pct': mean(100*int(r['actual'] > 0) for r in members)})
    answered = sum(r['actual'] is not None for r in selected)
    saved_days = len({r['target_date'] for r in rows})
    days = (end-start).days+1
    return {'start': start.isoformat(), 'end': end.isoformat(), 'calendar_days': days,
        'saved_days': saved_days, 'unrecorded_days': days-saved_days,
        'saved_subjects': len(rows), 'panel_saved_subjects': len(selected),
        'panel_share': len(selected)/len(rows) if rows else None,
        'resolved': answered, 'pending': len(selected)-answered,
        'resolution': answered/len(selected) if selected else None,
        'complete_days': len(daily), 'incomplete_days': saved_days-len(daily),
        'weeks': len({date.fromisoformat(r['date']).isocalendar()[:2] for r in daily}),
        'latest_complete_date': daily[-1]['date'] if daily else None,
        'metrics': {k: mean(r[k] for r in daily) if daily else None
                    for k in ('brier', 'mae_coins', 'avg_diff_coins', 'positive_pct')},
        'unresolved_reasons': dict(Counter(r['unresolved_reason'] or '結果待ち' for r in selected if r['actual'] is None)),
        'daily': daily}


def compare_cohort(rows, *, end):
    recent_start = end-timedelta(days=SPEC['recent_days']-1)
    baseline_end = recent_start-timedelta(days=1)
    start = recent_start-timedelta(days=SPEC['baseline_days'])
    rows = [r for r in rows if start.isoformat() <= r['target_date'] <= end.isoformat()]
    previous = [r for r in rows if r['target_date'] < recent_start.isoformat()]
    recent = [r for r in rows if r['target_date'] >= recent_start.isoformat()]
    panel = {tuple(r['key']) for r in previous} & {tuple(r['key']) for r in recent}
    # Fail closed if a subject/date was repeated, even with identical values.
    identities = [(r['target_date'], tuple(r['key'])) for r in rows]
    duplicate = len(identities) != len(set(identities))
    safe_rows = [] if duplicate else rows
    baseline = summarize_window(safe_rows, panel, start=start, end=baseline_end)
    latest = summarize_window(safe_rows, panel, start=recent_start, end=end)
    middle = recent_start+timedelta(days=14)
    halves = [summarize_window(safe_rows, panel, start=recent_start, end=middle-timedelta(days=1)),
              summarize_window(safe_rows, panel, start=middle, end=end)]
    reasons = []
    if duplicate:
        reasons.append('同じ対象・日の保存が重複しているため集計を保留')
    if not panel:
        reasons.append('両期間に共通する機種・台番号の事前記録がない')
    for label, window, minimum_days, minimum_weeks in (
            ('以前56日', baseline, SPEC['minimum_baseline_days'], SPEC['minimum_baseline_weeks']),
            ('直近28日', latest, SPEC['minimum_recent_days'], SPEC['minimum_recent_weeks'])):
        if window['complete_days'] < minimum_days or window['weeks'] < minimum_weeks:
            reasons.append(f'{label}の全件照合が{minimum_days}日・{minimum_weeks}週に不足')
        if window['resolution'] is None or window['resolution'] < SPEC['minimum_resolution']:
            reasons.append(f'{label}の共通対象の照合率が80%未満')
        if window['panel_share'] is None or window['panel_share'] < SPEC['minimum_panel_share']:
            reasons.append(f'{label}で共通対象が保存件数の80%未満（対象構成の変化）')
    if any(h['complete_days'] < SPEC['minimum_half_days'] for h in halves):
        reasons.append('直近の前半・後半それぞれに全件照合7日が必要')
    last = latest['latest_complete_date']
    if last is None or (end-date.fromisoformat(last)).days > SPEC['freshness_days']:
        reasons.append('直近7日内に共通対象全件の照合がない')
    b, r = baseline['metrics'], latest['metrics']
    delta = {k: r[k]-b[k] if r[k] is not None and b[k] is not None else None for k in b}
    alerts = []
    if not reasons:
        half_metrics = [h['metrics'] for h in halves]
        # Require the same signal in both recent halves; thresholds are fixed,
        # not tuned against the displayed outcomes. No statistical claim.
        if all(h['brier']-b['brier'] >= SPEC['brier_increase'] for h in half_metrics):
            alerts.append({'kind': 'forecast', 'metric': 'brier', 'message': '確率予測の誤差が直近の前半・後半とも拡大'})
        mae_limit = max(SPEC['mae_increase_coins'], b['mae_coins']*SPEC['mae_increase_ratio'])
        if all(h['mae_coins']-b['mae_coins'] >= mae_limit for h in half_metrics):
            alerts.append({'kind': 'forecast', 'metric': 'mae_coins', 'message': '差枚予測の誤差が直近の前半・後半とも拡大'})
        for key, threshold in [('avg_diff_coins', SPEC['outcome_shift_coins']), ('positive_pct', SPEC['positive_shift_points'])]:
            shifts = [h[key]-b[key] for h in half_metrics]
            if all(v >= threshold for v in shifts) or all(v <= -threshold for v in shifts):
                alerts.append({'kind': 'observed', 'metric': key,
                    'message': '共通対象の公開実績に同方向の変化。営業方針の変更と断定はできません'})
    return {'status': 'insufficient_data' if reasons else 'review' if alerts else 'no_alert',
        'reasons': reasons, 'alerts': alerts, 'common_subjects': len(panel),
        'baseline': baseline, 'recent': latest, 'halves': halves, 'delta': delta, 'live_applied': False}


def monitor_report(conn, *, as_of=None):
    as_of = stamp((as_of or datetime.now(JST)).isoformat())
    end = as_of.date()-timedelta(days=1)
    start = end-timedelta(days=SPEC['baseline_days']+SPEC['recent_days']-1)
    samples, excluded = read_samples(conn, since=start, as_of=as_of)
    groups = defaultdict(list)
    for r in samples:
        groups[r['version'], r['source_hash'], r['hall_name'], r['scope'], r['model']].append(r)
    items = []
    for (version, source, hall, scope, model), rows in sorted(groups.items()):
        result = compare_cohort(rows, end=end)
        if excluded['記録の整合性・保存時点違反'] or source.startswith('legacy:'):
            result['status'] = 'insufficient_data'
            result['alerts'] = []
            result['reasons'].append('保存記録の整合性または実装識別を確認できないため保留')
        items.append({'version': version, 'source_hash': source, 'hall_name': hall,
                      'scope': scope, 'model': model, 'label': MODELS[model], **result})
    recorded = len({r['target_date'] for r in samples})
    return {'protocol': PROTOCOL, 'spec': SPEC, 'as_of': as_of.isoformat(),
        'window_start': start.isoformat(), 'window_end': end.isoformat(),
        'recorded_days': recorded, 'unassessed_calendar_days': 84-recorded,
        'items': items, 'exclusions': dict(excluded), 'live_applied': False,
        'notice': '保存した予測の事後点検です。警告は統計的な確定判定ではなく、方式・順位・着席判定を自動変更しません。',
        'scope_notice': '四條畷周辺の同じ版・実装・店舗・方式・機種平均／台番号を分離。両期間共通の対象だけを比較し、全対象の結果が揃わない日は成績に含めません。',
        'bias_notice': '公開・稼働・イベント・対象構成の偏りは残ります。差枚の変化だけで設定変更や回収店と断定しません。警告なしも精度や安全の保証ではありません。'}

File: mobile/prediction-monitor.mjs
// Shared, read-only monitoring. No polling or automatic prediction changes.
const esc = v => String(v ?? '').replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
const num = (v, digits=1) => Number.isFinite(v) ? v.toFixed(digits) : '未計測';
const status = s => ({review:'要確認', insufficient_data:'データ不足・判定保留', no_alert:'固定基準の警告なし'})[s] || '未確認';

function comparison(item) {
  const metrics = [['brier','確率の誤差（小さい方が良い）',3], ['mae_coins','差枚の平均誤差（枚）',1],
    ['avg_diff_coins','共通対象の平均差枚（枚）',1], ['positive_pct','共通対象のプラス率（%）',1]];
  return `<div class="benchmark-table-scroll" tabindex="0" aria-label="以前と直近の成績・横スクロールできます"><table><caption>同じ機種・台番号で比較（日を等重み）</caption><thead><tr><th scope="col">指標</th><th scope="col">以前56日</th><th scope="col">直近28日</th><th scope="col">変化量</th></tr></thead><tbody>${metrics.map(([key,label,digits])=>`<tr><th scope="row">${label}</th><td>${num(item.baseline.metrics[key],digits)}</td><td>${num(item.recent.metrics[key],digits)}</td><td>${num(item.delta[key],digits)}</td></tr>`).join('')}</tbody></table></div>`;
}

function windowMarkup(w, label) {
  return `<p><b>${label}：${esc(w.start)}〜${esc(w.end)}</b><br>全件照合 ${esc(w.complete_days)}日 ／ 保存 ${esc(w.saved_days)}日 ／ 未保存 ${esc(w.unrecorded_days)}日<br>保存された日で全件未照合 ${esc(w.incomplete_days)}日 ／ 共通対象の結果待ち ${esc(w.pending)}件<br>共通対象の照合率 ${num(w.resolution==null?null:w.resolution*100)}% ／ 保存件数に占める共通対象 ${num(w.panel_share==null?null:w.panel_share*100)}%<br>最終全件照合 ${esc(w.latest_complete_date || 'なし')}</p><p>${Object.entries(w.unresolved_reasons || {}).map(([k,v])=>`${esc(k)} ${esc(v)}件`).join('／')}</p>`;
}

export function monitorMarkup(report) {
  if (!report) return '<p>監視結果をまだ取得できていません。</p>';
  const items=report.items || [];
  const halls=[...new Set(items.map(r=>r.hall_name))];
  return `<p>対象期間 ${esc(report.window_start)}〜${esc(report.window_end)}<br>検証可能な保存 ${esc(report.recorded_days)}日 ／ 未保存・対象外 ${esc(report.unassessed_calendar_days)}日</p>
    <p><b>要確認 ${items.filter(r=>r.status==='review').length}区分 ／ データ不足 ${items.filter(r=>r.status==='insufficient_data').length}区分 ／ 固定基準の警告なし ${items.filter(r=>r.status==='no_alert').length}区分</b></p>
    <p>${esc(report.notice)}</p>
    ${halls.map(hall=>`<details class="model-review-item"><summary>${esc(hall)}：${items.filter(r=>r.hall_name===hall && r.status==='review').length}区分が要確認</summary>
      ${items.filter(r=>r.hall_name===hall).map(r=>`<details><summary>${r.scope==='seat'?'台番号':'機種平均'}・${esc(r.label)} ／ ${status(r.status)} ／ v${esc(r.version)}</summary>
        <p>${(r.reasons || []).map(esc).join('／')}</p>
        ${(r.alerts || []).map(a=>`<p><b>${a.kind==='forecast'?'予測誤差':a.metric==='avg_diff_coins'?'平均差枚の変化':'プラス率の変化'}：</b>${esc(a.message)}</p>`).join('')}
        <p>共通対象 ${esc(r.common_subjects)}件。判定保留中の数値も参考表示であり、精度低下を確定したものではありません。</p>
        ${comparison(r)}
        <details><summary>照合状況・警告の根拠</summary>${windowMarkup(r.baseline,'以前56日')}${windowMarkup(r.recent,'直近28日')}
          <p>直近の前半／後半の全件照合：${(r.halves || []).map(h=>`${esc(h.complete_days)}日`).join(' ／ ')}。両方で同じ基準を超えた場合だけ警告します。</p>
          ${(r.halves || []).map((h,i)=>`<p>${i===0?'前半14日':'後半14日'}：確率誤差 ${num(h.metrics.brier,3)} ／ 差枚誤差 ${num(h.metrics.mae_coins)}枚 ／ 平均差枚 ${num(h.metrics.avg_diff_coins)}枚 ／ プラス率 ${num(h.metrics.positive_pct)}%</p>`).join('')}
          <small>実装識別 ${esc(r.source_hash)}</small></details></details>`).join('')}</details>`).join('') || '<p>比較できる事前記録がまだありません。予測悪化や「強い店がない」という意味ではありません。</p>'}
    <p>${esc(report.scope_notice)}</p><p>${esc(report.bias_notice)}</p>
    <p>${Object.entries(report.exclusions || {}).map(([k,v])=>`${esc(k)} ${esc(v)}日分`).join('／')}</p>`;
}

export async function mountPredictionMonitor(element, request) {
  element.innerHTML=`<details class="model-review prediction-monitor" open><summary>店舗傾向・予測の変化を点検（3.33）</summary>
    <p>「予測が外れやすくなった？」と「公開実績が変わった？」を分けて確認します。推奨候補だけでなく、参考・見送りを含む保存対象の点検です。</p>
    <div data-monitor-body role="status">保存記録を確認中…</div><button type="button" data-monitor-refresh>変化を再確認</button><p data-monitor-message role="status"></p>
    <details><summary>比較期間と警告の条件</summary><p>昨日までの直近28日と、その前56日。以前は全件照合28日・6週、直近は14日・3週かつ前半／後半各7日、照合率・共通対象割合とも80%以上、7日以内の全件照合が必要です。</p>
    <p>直近の前半／後半がともに以前より、確率誤差0.05以上、または差枚誤差100枚・20%の大きい方以上悪化したら要確認。公開実績は平均差枚500枚以上、またはプラス率20ポイント以上の同方向変化を別表示します。</p>
    <p>確率誤差は予測確率と実際のプラス／非プラスのずれです。基準は暫定の運用目安で、統計的な確定判定や利益保証ではありません。これは画面を開いた時・再確認時の点検で、常時通知ではありません。</p></details></details>`;
  const body=element.querySelector('[data-monitor-body]'), button=element.querySelector('[data-monitor-refresh]'), message=element.querySelector('[data-monitor-message]');
  let busy=false;
  async function refresh() {
    if (busy) return;
    busy=true; button.disabled=true;
    try {
      const report=await request('/api/predictions/monitor');
      if (!element.isConnected) return;
      body.innerHTML=monitorMarkup(report);
      message.textContent='保存記録を点検しました。予測や着席判定は変更していません。';
    } catch (error) {
      if (!element.isConnected) return;
      body.textContent='監視結果を取得できませんでした。';
      message.textContent=`通信とサーバーの更新を確認して再確認してください：${error.message}`;
    } finally {busy=false; button.disabled=false;}
  }
  button.onclick=refresh;
  await refresh();
}
